### ISSUE TITLE
Reuse model.fit may overestimate the performance in cross-validation

### ISSUE THREAD

**Issue by zealseeker** [deepchem/deepchem#926]
I was trying to evaluate a model via cross-validation,
here is my code:
```python
def tf_graph_model(dataset, transformers=None, testset=None):
    # Batch size of models
    batch_size = 50
    model = GraphConvTensorGraph(
        1, batch_size=batch_size, mode='classification')
    spliter = dc.splits.RandomSplitter()
    split_datas = spliter.k_fold_split(dataset,5)
    scores = {m.name:[] for m in metrics}
    for train_set, val_set in split_datas:
        # Fit trained model
        model.fit(train_set, nb_epoch=10)
        valid_scores = model.evaluate(val_set,metrics,transformers)
        for name in valid_scores:
            scores[name].append(valid_scores[name])
    for name in scores:
        scores[name] = np.mean(scores[name])
    return scores
```

In this function, I created only one model instance and fed it repeatedly during 5-fold cross-validation.
I found that the performance of the model will be overestimated according to the result that the scores (metrics) would keep increasing in cross-validation (i.e. the performance of the 2nd fold was better then the 1st; the 3rd was better than the 1st and the 2nd ...). So the mean of the five scores cannot represent the real performance of the model.

I guess that when the model was fed by another training set, the variables/weights of the model were not initialed again, instead, the optimization was based on the previous fed model.

An alternative way is to create a new model instance for each split data. But Is there any way better like 
 `model.init()`

**Comment by rbharath** [comment]
Each split should have its own model object. The current design leaks information since weights are shared. For example:

```
def tf_graph_model(dataset, transformers=None, testset=None):
    # Batch size of models
    batch_size = 50
    spliter = dc.splits.RandomSplitter()
    split_datas = spliter.k_fold_split(dataset,5)
    scores = {m.name:[] for m in metrics}
    for train_set, val_set in split_datas:
        model = GraphConvTensorGraph(
            1, batch_size=batch_size, mode='classification')
        # Fit trained model
        model.fit(train_set, nb_epoch=10)
        valid_scores = model.evaluate(val_set,metrics,transformers)
        for name in valid_scores:
            scores[name].append(valid_scores[name])
    for name in scores:
        scores[name] = np.mean(scores[name])
    return scores
```

**Comment by SeanXiaoyuSun** [comment]
@rbharath Sorry for this novice question. If we want to conduct cross validation, why should we give every split a model object? Shouldn't we train a single model with these 5 train/val splits and test this model by using an additional test set?

**Comment by peastman** [comment]
That's not what cross validation is.  In cross validation, you divide your dataset into N parts.  You then train N models, each time using N-1 of those parts as the training set and the final part as the test set.  So in this case, you would train the model on 4/5 of the data then test it on the remaining 1/5.  And you would do that 5 times, each time using a different 1/5 of the data as the test set.
